exist-core/src/main/java/org/exist/util/io/VirtualTempPath.java

```diff
@@ -41,73 +42,102 @@
     private static final byte[] EMPTY_BUFFER = new byte[0];
 
     private final int inMemorySize;
+    private final StampedLock lock;
     private final TemporaryFileManager tempFileManager;
 
     private MemoryContents content;
     private Path contentFile;
 
     public VirtualTempPath(int inMemorySize, TemporaryFileManager tempFileManager) {
-        this.tempFileManager = tempFileManager;
         this.inMemorySize = inMemorySize;
+        this.lock = new StampedLock();
+        this.tempFileManager = tempFileManager;
     }
 
     private OutputStream initOverflowOutputStream() throws IOException {
-        if (contentFile == null) {
-            contentFile = tempFileManager.getTemporaryFile();
-        }
-        if (LOG.isDebugEnabled()) {
-            LOG.debug("Initializing overflow to " + contentFile.toAbsolutePath());
+        long stamp = lock.writeLock();
+        try {
+            if (contentFile == null) {
+                contentFile = tempFileManager.getTemporaryFile();
+            }
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Initializing overflow to " + contentFile.toAbsolutePath());
+            }
+            return Files.newOutputStream(contentFile);
+
+        } finally {
+            lock.unlockWrite(stamp);
         }
-        return Files.newOutputStream(contentFile);
     }
 
     public OutputStream newOutputStream() throws IOException {
-        if (inMemorySize <= 0) {
-            contentFile = tempFileManager.getTemporaryFile();
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("In memory buffering disabled writing to " + contentFile.toAbsolutePath());
+        long stamp = lock.readLock();
+        try {
+            if (inMemorySize <= 0) {
+                contentFile = tempFileManager.getTemporaryFile();
```

Comment:
You have a write to `contentFile` here happening under a read lock!?!

It might be useful to add `@GuardedBy` annotations to `content` and `contentFile` members so that FindBugs plugin will show you such issues ;-)